After three atempts, I've finally found a suitable dataset!

This first dataset that I chose turned out to be bogus - it was synthetically created and didn't give any information about where it came from.

The second dataset came from an identified source; however, it was too simple. Incredibly easy to train models with almost perfect accuracy.

This third dataset comes from a recognised source [1]. This dataset was more challenging and allowed me to use some of the methods learned on the Professional Certificate course.

##### [1] Davide Anguita, Alessandro Ghio, Luca Oneto, Xavier Parra and Jorge L. Reyes-Ortiz. A Public Domain Dataset for Human Activity Recognition Using Smartphones. 21st European Symposium on Artificial Neural Networks, Computational Intelligence and Machine Learning, ESANN 2013. Bruges, Belgium 24-26 April 2013.


https://www.kaggle.com/code/levkharlashkin/human-activity-recognition?select=test.csv

In [961]:
# The first step is to import some libraries

import numpy as np # linear algebra
import pandas as pd # data processing

from sklearn import metrics #Import scikit-learn metrics module for accuracy calculation

from sklearn.metrics import classification_report

In [963]:
# I now read the csv file and make a copy
X_train = pd.read_csv("human_activity/train.csv")
X_test = pd.read_csv("human_activity/test.csv")

In [964]:
# Examining the dataframe...
X_train

,tBodyAcc-mean()-X,tBodyAcc-mean()-Y,tBodyAcc-mean()-Z,tBodyAcc-std()-X,tBodyAcc-std()-Y,tBodyAcc-std()-Z,tBodyAcc-mad()-X,tBodyAcc-mad()-Y,tBodyAcc-mad()-Z,tBodyAcc-max()-X,...,fBodyBodyGyroJerkMag-kurtosis(),"angle(tBodyAccMean,gravity)","angle(tBodyAccJerkMean),gravityMean)","angle(tBodyGyroMean,gravityMean)","angle(tBodyGyroJerkMean,gravityMean)","angle(X,gravityMean)","angle(Y,gravityMean)","angle(Z,gravityMean)",subject,Activity
0,0.288585,-0.020294,-0.132905,-0.995279,-0.983111,-0.913526,-0.995112,-0.983185,-0.923527,-0.934724,...,-0.710304,-0.112754,0.030400,-0.464761,-0.018446,-0.841247,0.179941,-0.058627,1,STANDING
1,0.278419,-0.016411,-0.123520,-0.998245,-0.975300,-0.960322,-0.998807,-0.974914,-0.957686,-0.943068,...,-0.861499,0.053477,-0.007435,-0.732626,0.703511,-0.844788,0.180289,-0.054317,1,STANDING
2,0.279653,-0.019467,-0.113462,-0.995380,-0.967187,-0.978944,-0.996520,-0.963668,-0.977469,-0.938692,...,-0.760104,-0.118559,0.177899,0.100699,0.808529,-0.848933,0.180637,-0.049118,1,STANDING
3,0.279174,-0.026201,-0.123283,-0.996091,-0.983403,-0.990675,-0.997099,-0.982750,-0.989302,-0.938692,...,-0.482845,-0.036788,-0.012892,0.640011,-0.485366,-0.848649,0.181935,-0.047663,1,STANDING
4,0.276629,-0.016570,-0.115362,-0.998139,-0.980817,-0.990482,-0.998321,-0.979672,-0.990441,-0.942469,...,-0.699205,0.123320,0.122542,0.693578,-0.615971,-0.847865,0.185151,-0.043892,1,STANDING
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7347,0.299665,-0.057193,-0.181233,-0.195387,0.039905,0.077078,-0.282301,0.043616,0.060410,0.210795,...,-0.880324,-0.190437,0.829718,0.206972,-0.425619,-0.791883,0.238604,0.049819,30,WALKING_UPSTAIRS
7348,0.273853,-0.007749,-0.147468,-0.235309,0.004816,0.059280,-0.322552,-0.029456,0.080585,0.117440,...,-0.680744,0.064907,0.875679,-0.879033,0.400219,-0.771840,0.252676,0.050053,30,WALKING_UPSTAIRS
7349,0.273387,-0.017011,-0.045022,-0.218218,-0.103822,0.274533,-0.304515,-0.098913,0.332584,0.043999,...,-0.304029,0.052806,-0.266724,0.864404,0.701169,-0.779133,0.249145,0.040811,30,WALKING_UPSTAIRS
7350,0.289654,-0.018843,-0.158281,-0.219139,-0.111412,0.268893,-0.310487,-0.068200,0.319473,0.101702,...,-0.344314,-0.101360,0.700740,0.936674,-0.589479,-0.785181,0.246432,0.025339,30,WALKING_UPSTAIRS


In [965]:
X_test

,tBodyAcc-mean()-X,tBodyAcc-mean()-Y,tBodyAcc-mean()-Z,tBodyAcc-std()-X,tBodyAcc-std()-Y,tBodyAcc-std()-Z,tBodyAcc-mad()-X,tBodyAcc-mad()-Y,tBodyAcc-mad()-Z,tBodyAcc-max()-X,...,fBodyBodyGyroJerkMag-kurtosis(),"angle(tBodyAccMean,gravity)","angle(tBodyAccJerkMean),gravityMean)","angle(tBodyGyroMean,gravityMean)","angle(tBodyGyroJerkMean,gravityMean)","angle(X,gravityMean)","angle(Y,gravityMean)","angle(Z,gravityMean)",subject,Activity
0,0.257178,-0.023285,-0.014654,-0.938404,-0.920091,-0.667683,-0.952501,-0.925249,-0.674302,-0.894088,...,-0.705974,0.006462,0.162920,-0.825886,0.271151,-0.720009,0.276801,-0.057978,2,STANDING
1,0.286027,-0.013163,-0.119083,-0.975415,-0.967458,-0.944958,-0.986799,-0.968401,-0.945823,-0.894088,...,-0.594944,-0.083495,0.017500,-0.434375,0.920593,-0.698091,0.281343,-0.083898,2,STANDING
2,0.275485,-0.026050,-0.118152,-0.993819,-0.969926,-0.962748,-0.994403,-0.970735,-0.963483,-0.939260,...,-0.640736,-0.034956,0.202302,0.064103,0.145068,-0.702771,0.280083,-0.079346,2,STANDING
3,0.270298,-0.032614,-0.117520,-0.994743,-0.973268,-0.967091,-0.995274,-0.974471,-0.968897,-0.938610,...,-0.736124,-0.017067,0.154438,0.340134,0.296407,-0.698954,0.284114,-0.077108,2,STANDING
4,0.274833,-0.027848,-0.129527,-0.993852,-0.967445,-0.978295,-0.994111,-0.965953,-0.977346,-0.938610,...,-0.846595,-0.002223,-0.040046,0.736715,-0.118545,-0.692245,0.290722,-0.073857,2,STANDING
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2942,0.310155,-0.053391,-0.099109,-0.287866,-0.140589,-0.215088,-0.356083,-0.148775,-0.232057,0.185361,...,-0.750809,-0.337422,0.346295,0.884904,-0.698885,-0.651732,0.274627,0.184784,24,WALKING_UPSTAIRS
2943,0.363385,-0.039214,-0.105915,-0.305388,0.028148,-0.196373,-0.373540,-0.030036,-0.270237,0.185361,...,-0.700274,-0.736701,-0.372889,-0.657421,0.322549,-0.655181,0.273578,0.182412,24,WALKING_UPSTAIRS
2944,0.349966,0.030077,-0.115788,-0.329638,-0.042143,-0.250181,-0.388017,-0.133257,-0.347029,0.007471,...,-0.467179,-0.181560,0.088574,0.696663,0.363139,-0.655357,0.274479,0.181184,24,WALKING_UPSTAIRS
2945,0.237594,0.018467,-0.096499,-0.323114,-0.229775,-0.207574,-0.392380,-0.279610,-0.289477,0.007471,...,-0.617737,0.444558,-0.819188,0.929294,-0.008398,-0.659719,0.264782,0.187563,24,WALKING_UPSTAIRS


In [966]:
y_train = X_train.Activity

In [971]:
X_train = X_train.drop("Activity", axis='columns')

In [972]:
X_train

,tBodyAcc-mean()-X,tBodyAcc-mean()-Y,tBodyAcc-mean()-Z,tBodyAcc-std()-X,tBodyAcc-std()-Y,tBodyAcc-std()-Z,tBodyAcc-mad()-X,tBodyAcc-mad()-Y,tBodyAcc-mad()-Z,tBodyAcc-max()-X,...,fBodyBodyGyroJerkMag-skewness(),fBodyBodyGyroJerkMag-kurtosis(),"angle(tBodyAccMean,gravity)","angle(tBodyAccJerkMean),gravityMean)","angle(tBodyGyroMean,gravityMean)","angle(tBodyGyroJerkMean,gravityMean)","angle(X,gravityMean)","angle(Y,gravityMean)","angle(Z,gravityMean)",subject
0,0.288585,-0.020294,-0.132905,-0.995279,-0.983111,-0.913526,-0.995112,-0.983185,-0.923527,-0.934724,...,-0.298676,-0.710304,-0.112754,0.030400,-0.464761,-0.018446,-0.841247,0.179941,-0.058627,1
1,0.278419,-0.016411,-0.123520,-0.998245,-0.975300,-0.960322,-0.998807,-0.974914,-0.957686,-0.943068,...,-0.595051,-0.861499,0.053477,-0.007435,-0.732626,0.703511,-0.844788,0.180289,-0.054317,1
2,0.279653,-0.019467,-0.113462,-0.995380,-0.967187,-0.978944,-0.996520,-0.963668,-0.977469,-0.938692,...,-0.390748,-0.760104,-0.118559,0.177899,0.100699,0.808529,-0.848933,0.180637,-0.049118,1
3,0.279174,-0.026201,-0.123283,-0.996091,-0.983403,-0.990675,-0.997099,-0.982750,-0.989302,-0.938692,...,-0.117290,-0.482845,-0.036788,-0.012892,0.640011,-0.485366,-0.848649,0.181935,-0.047663,1
4,0.276629,-0.016570,-0.115362,-0.998139,-0.980817,-0.990482,-0.998321,-0.979672,-0.990441,-0.942469,...,-0.351471,-0.699205,0.123320,0.122542,0.693578,-0.615971,-0.847865,0.185151,-0.043892,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7347,0.299665,-0.057193,-0.181233,-0.195387,0.039905,0.077078,-0.282301,0.043616,0.060410,0.210795,...,-0.588433,-0.880324,-0.190437,0.829718,0.206972,-0.425619,-0.791883,0.238604,0.049819,30
7348,0.273853,-0.007749,-0.147468,-0.235309,0.004816,0.059280,-0.322552,-0.029456,0.080585,0.117440,...,-0.390738,-0.680744,0.064907,0.875679,-0.879033,0.400219,-0.771840,0.252676,0.050053,30
7349,0.273387,-0.017011,-0.045022,-0.218218,-0.103822,0.274533,-0.304515,-0.098913,0.332584,0.043999,...,0.025145,-0.304029,0.052806,-0.266724,0.864404,0.701169,-0.779133,0.249145,0.040811,30
7350,0.289654,-0.018843,-0.158281,-0.219139,-0.111412,0.268893,-0.310487,-0.068200,0.319473,0.101702,...,0.063907,-0.344314,-0.101360,0.700740,0.936674,-0.589479,-0.785181,0.246432,0.025339,30


In [975]:
y_train

0               STANDING
1               STANDING
2               STANDING
3               STANDING
4               STANDING
              ...       
7347    WALKING_UPSTAIRS
7348    WALKING_UPSTAIRS
7349    WALKING_UPSTAIRS
7350    WALKING_UPSTAIRS
7351    WALKING_UPSTAIRS
Name: Activity, Length: 7352, dtype: object

In [977]:
# Any missing data?
X_train.isnull().sum()

tBodyAcc-mean()-X                       0
tBodyAcc-mean()-Y                       0
tBodyAcc-mean()-Z                       0
tBodyAcc-std()-X                        0
tBodyAcc-std()-Y                        0
                                       ..
angle(tBodyGyroJerkMean,gravityMean)    0
angle(X,gravityMean)                    0
angle(Y,gravityMean)                    0
angle(Z,gravityMean)                    0
subject                                 0
Length: 562, dtype: int64

In [979]:
y_test = X_test.Activity

In [981]:
X_test = X_test.drop("Activity", axis='columns')

In [983]:
X_test

,tBodyAcc-mean()-X,tBodyAcc-mean()-Y,tBodyAcc-mean()-Z,tBodyAcc-std()-X,tBodyAcc-std()-Y,tBodyAcc-std()-Z,tBodyAcc-mad()-X,tBodyAcc-mad()-Y,tBodyAcc-mad()-Z,tBodyAcc-max()-X,...,fBodyBodyGyroJerkMag-skewness(),fBodyBodyGyroJerkMag-kurtosis(),"angle(tBodyAccMean,gravity)","angle(tBodyAccJerkMean),gravityMean)","angle(tBodyGyroMean,gravityMean)","angle(tBodyGyroJerkMean,gravityMean)","angle(X,gravityMean)","angle(Y,gravityMean)","angle(Z,gravityMean)",subject
0,0.257178,-0.023285,-0.014654,-0.938404,-0.920091,-0.667683,-0.952501,-0.925249,-0.674302,-0.894088,...,-0.330370,-0.705974,0.006462,0.162920,-0.825886,0.271151,-0.720009,0.276801,-0.057978,2
1,0.286027,-0.013163,-0.119083,-0.975415,-0.967458,-0.944958,-0.986799,-0.968401,-0.945823,-0.894088,...,-0.121845,-0.594944,-0.083495,0.017500,-0.434375,0.920593,-0.698091,0.281343,-0.083898,2
2,0.275485,-0.026050,-0.118152,-0.993819,-0.969926,-0.962748,-0.994403,-0.970735,-0.963483,-0.939260,...,-0.190422,-0.640736,-0.034956,0.202302,0.064103,0.145068,-0.702771,0.280083,-0.079346,2
3,0.270298,-0.032614,-0.117520,-0.994743,-0.973268,-0.967091,-0.995274,-0.974471,-0.968897,-0.938610,...,-0.344418,-0.736124,-0.017067,0.154438,0.340134,0.296407,-0.698954,0.284114,-0.077108,2
4,0.274833,-0.027848,-0.129527,-0.993852,-0.967445,-0.978295,-0.994111,-0.965953,-0.977346,-0.938610,...,-0.534685,-0.846595,-0.002223,-0.040046,0.736715,-0.118545,-0.692245,0.290722,-0.073857,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2942,0.310155,-0.053391,-0.099109,-0.287866,-0.140589,-0.215088,-0.356083,-0.148775,-0.232057,0.185361,...,-0.376278,-0.750809,-0.337422,0.346295,0.884904,-0.698885,-0.651732,0.274627,0.184784,24
2943,0.363385,-0.039214,-0.105915,-0.305388,0.028148,-0.196373,-0.373540,-0.030036,-0.270237,0.185361,...,-0.320418,-0.700274,-0.736701,-0.372889,-0.657421,0.322549,-0.655181,0.273578,0.182412,24
2944,0.349966,0.030077,-0.115788,-0.329638,-0.042143,-0.250181,-0.388017,-0.133257,-0.347029,0.007471,...,-0.118854,-0.467179,-0.181560,0.088574,0.696663,0.363139,-0.655357,0.274479,0.181184,24
2945,0.237594,0.018467,-0.096499,-0.323114,-0.229775,-0.207574,-0.392380,-0.279610,-0.289477,0.007471,...,-0.205445,-0.617737,0.444558,-0.819188,0.929294,-0.008398,-0.659719,0.264782,0.187563,24


In [985]:
y_test

0               STANDING
1               STANDING
2               STANDING
3               STANDING
4               STANDING
              ...       
2942    WALKING_UPSTAIRS
2943    WALKING_UPSTAIRS
2944    WALKING_UPSTAIRS
2945    WALKING_UPSTAIRS
2946    WALKING_UPSTAIRS
Name: Activity, Length: 2947, dtype: object

In [987]:
# Any missing data?
X_test.isnull().sum()

tBodyAcc-mean()-X                       0
tBodyAcc-mean()-Y                       0
tBodyAcc-mean()-Z                       0
tBodyAcc-std()-X                        0
tBodyAcc-std()-Y                        0
                                       ..
angle(tBodyGyroJerkMean,gravityMean)    0
angle(X,gravityMean)                    0
angle(Y,gravityMean)                    0
angle(Z,gravityMean)                    0
subject                                 0
Length: 562, dtype: int64

In [990]:
from sklearn import preprocessing

min_max_scaler = preprocessing.MinMaxScaler()
X_train = min_max_scaler.fit_transform(X_train)
X_train = pd.DataFrame(X_train)

X_test = min_max_scaler.fit_transform(X_test)
X_test = pd.DataFrame(X_test)

In [992]:
X_train

,0,1,2,3,4,5,6,7,8,9,...,552,553,554,555,556,557,558,559,560,561
0,0.644292,0.489853,0.433547,0.002361,0.008748,0.043237,0.002444,0.008546,0.038236,0.032638,...,0.350991,0.147940,0.437030,0.515200,0.267793,0.491741,0.079377,0.798251,0.470687,0.0
1,0.639209,0.491795,0.438240,0.000877,0.012824,0.019839,0.000596,0.012749,0.021157,0.028466,...,0.201676,0.070666,0.521131,0.496283,0.133774,0.853429,0.077606,0.798487,0.472842,0.0
2,0.639827,0.490266,0.443269,0.002310,0.017058,0.010528,0.001740,0.018464,0.011266,0.030654,...,0.304605,0.122488,0.434094,0.588950,0.550707,0.906041,0.075533,0.798722,0.475441,0.0
3,0.639587,0.486900,0.438359,0.001954,0.008596,0.004662,0.001450,0.008767,0.005349,0.030654,...,0.442374,0.264192,0.475464,0.493554,0.820538,0.257822,0.075675,0.799600,0.476168,0.0
4,0.638314,0.491715,0.442319,0.000931,0.009945,0.004759,0.000839,0.010331,0.004779,0.028765,...,0.324393,0.153612,0.556466,0.561271,0.847339,0.192392,0.076067,0.801776,0.478054,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7347,0.649833,0.471403,0.409383,0.402307,0.542650,0.538539,0.358850,0.530383,0.530205,0.605398,...,0.205011,0.061044,0.397729,0.914859,0.603878,0.287755,0.104058,0.837938,0.524910,1.0
7348,0.636926,0.496125,0.426266,0.382346,0.524338,0.529640,0.338724,0.493247,0.540293,0.558720,...,0.304610,0.163047,0.526914,0.937840,0.060523,0.701485,0.114080,0.847458,0.525026,1.0
7349,0.636694,0.491495,0.477489,0.390891,0.467640,0.637266,0.347742,0.457948,0.666292,0.521999,...,0.514134,0.355582,0.520791,0.366638,0.932807,0.852256,0.110434,0.845069,0.520406,1.0
7350,0.644827,0.490578,0.420860,0.390430,0.463679,0.634447,0.344756,0.473556,0.659737,0.550851,...,0.533662,0.334993,0.442795,0.850370,0.968966,0.205664,0.107409,0.843234,0.512670,1.0


In [994]:
X_test

,0,1,2,3,4,5,6,7,8,9,...,552,553,554,555,556,557,558,559,560,561
0,0.671879,0.557643,0.524648,0.041779,0.039955,0.222530,0.032601,0.037334,0.227491,0.033512,...,0.334815,0.147013,0.503509,0.584075,0.086554,0.633946,0.145366,0.622095,0.463627,0.0
1,0.694704,0.574264,0.427079,0.016514,0.016271,0.036272,0.008768,0.015757,0.037197,0.033512,...,0.439077,0.202528,0.458505,0.510621,0.282417,0.960119,0.157427,0.624468,0.450144,0.0
2,0.686363,0.553102,0.427948,0.003951,0.015037,0.024322,0.003484,0.014590,0.024821,0.007532,...,0.404789,0.179632,0.482788,0.603967,0.531794,0.570623,0.154851,0.623810,0.452512,0.0
3,0.682260,0.542325,0.428538,0.003320,0.013366,0.021404,0.002879,0.012722,0.021026,0.007906,...,0.327791,0.131938,0.491737,0.579791,0.669885,0.646630,0.156952,0.625916,0.453676,0.0
4,0.685848,0.550151,0.417320,0.003928,0.016277,0.013878,0.003687,0.016981,0.015105,0.007906,...,0.232658,0.076702,0.499164,0.481554,0.868285,0.438227,0.160644,0.629369,0.455367,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2942,0.713794,0.508207,0.445740,0.485861,0.429705,0.526560,0.447047,0.425588,0.537434,0.654315,...,0.311861,0.124595,0.331472,0.676700,0.942420,0.146759,0.182935,0.620958,0.589912,1.0
2943,0.755910,0.531487,0.439381,0.473900,0.514074,0.539131,0.434917,0.484960,0.510676,0.654315,...,0.339791,0.149863,0.131722,0.313430,0.170833,0.659759,0.181038,0.620411,0.588678,1.0
2944,0.745294,0.645268,0.430157,0.457346,0.478929,0.502986,0.424857,0.433347,0.456857,0.552008,...,0.440573,0.266410,0.409445,0.546522,0.848248,0.680145,0.180941,0.620881,0.588039,1.0
2945,0.656384,0.626202,0.448179,0.461799,0.385112,0.531607,0.421825,0.360167,0.497192,0.552008,...,0.397277,0.191132,0.722677,0.087998,0.964628,0.493546,0.178541,0.615814,0.591358,1.0


In [996]:
y_train.to_frame()
y_test.to_frame()

,Activity
0,STANDING
1,STANDING
2,STANDING
3,STANDING
4,STANDING
...,...
2942,WALKING_UPSTAIRS
2943,WALKING_UPSTAIRS
2944,WALKING_UPSTAIRS
2945,WALKING_UPSTAIRS


In [998]:
y_train = y_train.reset_index()
y_test = y_test.reset_index()
y_train = y_train.drop("index", axis='columns')
y_test = y_test.drop("index", axis='columns')

In [1000]:
y_test

,Activity
0,STANDING
1,STANDING
2,STANDING
3,STANDING
4,STANDING
...,...
2942,WALKING_UPSTAIRS
2943,WALKING_UPSTAIRS
2944,WALKING_UPSTAIRS
2945,WALKING_UPSTAIRS


In [1002]:
# Find the unique numbers from the train labels
classes = np.unique(y_train)
nClasses = len(classes)
print('Total number of outputs : ', nClasses)
print('Output classes : ', classes)

Total number of outputs :  6
Output classes :  ['LAYING' 'SITTING' 'STANDING' 'WALKING' 'WALKING_DOWNSTAIRS'
 'WALKING_UPSTAIRS']


In [1004]:
target_names = ['Walking', 'Walking Upstairs', 'Walking Downstairs', 'Sitting', 'Standing', 'Laying']

In [ ]:
print('Training data shape:', X_train.shape, y_train.shape)
print('Testing data shape:', X_test.shape, y_test.shape)

# PCA

In [1122]:
#Dimensionality Reduction
from sklearn.decomposition import PCA

pca = PCA(n_components=40)
X_train_pc = pca.fit_transform(X_train)
X_test_pc = pca.transform(X_test)
explained_variance = pca.explained_variance_ratio_
print(pd.DataFrame(explained_variance))

print(sum(pca.explained_variance_ratio_))

           0
0   0.615828
1   0.050640
2   0.042643
3   0.018626
4   0.016604
5   0.012495
6   0.011697
7   0.010868
8   0.009834
9   0.008507
10  0.008370
11  0.007051
12  0.006703
13  0.005993
14  0.005412
15  0.004889
16  0.004738
17  0.004560
18  0.004240
19  0.004112
20  0.004092
21  0.003869
22  0.003706
23  0.003532
24  0.003417
25  0.003249
26  0.003109
27  0.003044
28  0.002983
29  0.002885
30  0.002804
31  0.002705
32  0.002627
33  0.002561
34  0.002387
35  0.002372
36  0.002251
37  0.002177
38  0.002078
39  0.002057
0.9117156861660729


# MODEL TRAINING

# Decision Tree

In [1126]:
# Load libraries
from sklearn.tree import DecisionTreeClassifier # Import Decision Tree Classifier

# Create Decision Tree classifer object
clf = DecisionTreeClassifier(random_state=16)

# Train Decision Tree Classifer
clf = clf.fit(X_train_pc,y_train)

#Predict the response for test dataset
y_pred = clf.predict(X_test_pc)

# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

print(classification_report(y_test,y_pred, target_names=target_names))

Accuracy: 0.7343060739735324
                    precision    recall  f1-score   support

           Walking       0.98      0.98      0.98       537
  Walking Upstairs       0.64      0.79      0.71       491
Walking Downstairs       0.77      0.61      0.68       532
           Sitting       0.80      0.58      0.67       496
          Standing       0.53      0.84      0.65       420
            Laying       0.80      0.60      0.68       471

          accuracy                           0.73      2947
         macro avg       0.75      0.73      0.73      2947
      weighted avg       0.76      0.73      0.74      2947



## Decision Tree hyperparameter tuning: Grid Search

In [1129]:
# Create the parameter grid based on the results of random search 
params = {
    'max_depth': [2, 3, 4, 5, 6, 7, 8, 9, 10, 15, 20, 25],
    'min_samples_leaf': [5, 6, 7, 8, 9, 10, 15, 20, 25],
    'criterion': ["gini", "entropy"]
}

from sklearn.model_selection import GridSearchCV

# Instantiate the grid search model
grid_search = GridSearchCV(estimator=clf, 
                           param_grid=params, 
                           cv=4, n_jobs=-1, verbose=1, scoring = "accuracy")

grid_search.fit(X_train_pc, y_train)

Fitting 4 folds for each of 216 candidates, totalling 864 fits


GridSearchCV(cv=4, estimator=DecisionTreeClassifier(random_state=16), n_jobs=-1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [2, 3, 4, 5, 6, 7, 8, 9, 10, 15, 20, 25],
                         'min_samples_leaf': [5, 6, 7, 8, 9, 10, 15, 20, 25]},
             scoring='accuracy', verbose=1)

In [1130]:
score_df = pd.DataFrame(grid_search.cv_results_)

In [1131]:
grid_search.best_estimator_

DecisionTreeClassifier(criterion='entropy', max_depth=7, min_samples_leaf=20,
                       random_state=16)

In [1132]:
grid_search.best_score_

0.7861806311207835

In [1192]:
# Create Decision Tree classifer object
clf = DecisionTreeClassifier(criterion='entropy', max_depth=7, min_samples_leaf=20, random_state=16)

# Train Decision Tree Classifer
clf = clf.fit(X_train_pc,y_train)

#Predict the response for test dataset
y_pred = clf.predict(X_test_pc)

# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

print(classification_report(y_test,y_pred, target_names=target_names))

Accuracy: 0.7495758398371225
                    precision    recall  f1-score   support

           Walking       0.98      0.96      0.97       537
  Walking Upstairs       0.69      0.77      0.73       491
Walking Downstairs       0.79      0.71      0.75       532
           Sitting       0.81      0.54      0.65       496
          Standing       0.52      0.93      0.67       420
            Laying       0.87      0.60      0.71       471

          accuracy                           0.75      2947
         macro avg       0.78      0.75      0.75      2947
      weighted avg       0.79      0.75      0.75      2947



# Random Forest Classifier

In [1140]:
from sklearn.ensemble import RandomForestClassifier

# Create RF classifer object
clf = RandomForestClassifier(random_state=123)

# Train RF Classifer
clf = clf.fit(X_train_pc,y_train)

#Predict the response for test dataset
y_pred = clf.predict(X_test_pc)

# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

print(classification_report(y_test,y_pred, target_names=target_names))

Accuracy: 0.827281981676281
                    precision    recall  f1-score   support

           Walking       0.99      0.99      0.99       537
  Walking Upstairs       0.81      0.89      0.85       491
Walking Downstairs       0.90      0.81      0.85       532
           Sitting       0.98      0.63      0.77       496
          Standing       0.55      0.95      0.70       420
            Laying       0.94      0.69      0.80       471

          accuracy                           0.83      2947
         macro avg       0.86      0.83      0.83      2947
      weighted avg       0.87      0.83      0.83      2947



## Random Forest hyperparameter tuning: Grid Search

In [1142]:
params = {
    'n_estimators': [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
    'max_depth': [2,3,4,5,6,7,8,9,10,11,12,13,14,15]
}

# Create a random forest classifier
rf = RandomForestClassifier()

# Instantiate the grid search model
grid_search = GridSearchCV(estimator=rf, 
                           param_grid=params, 
                           cv=4, n_jobs=-1, verbose=1, scoring = "accuracy")

# Fit the random search object to the data
grid_search.fit(X_train_pc, y_train)

Fitting 4 folds for each of 210 candidates, totalling 840 fits


GridSearchCV(cv=4, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'max_depth': [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,
                                       14, 15],
                         'n_estimators': [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
                                          16, 17, 18, 19]},
             scoring='accuracy', verbose=1)

In [1143]:
grid_search.best_estimator_

RandomForestClassifier(max_depth=12, n_estimators=18)

In [1144]:
grid_search.best_score_

0.8477965179542981

In [1145]:
print(classification_report(y_test,y_pred, target_names=target_names))

                    precision    recall  f1-score   support

           Walking       0.99      0.99      0.99       537
  Walking Upstairs       0.81      0.89      0.85       491
Walking Downstairs       0.90      0.81      0.85       532
           Sitting       0.98      0.63      0.77       496
          Standing       0.55      0.95      0.70       420
            Laying       0.94      0.69      0.80       471

          accuracy                           0.83      2947
         macro avg       0.86      0.83      0.83      2947
      weighted avg       0.87      0.83      0.83      2947



# Logistic Regression

In [1151]:
# import the class
from sklearn.linear_model import LogisticRegression

# instantiate the model (using the default parameters)
logreg = LogisticRegression(random_state=16)

# fit the model with data
logreg.fit(X_train, y_train)

y_pred = logreg.predict(X_test)

print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.8476416694944011


In [1153]:
print(classification_report(y_test,y_pred, target_names=target_names))

                    precision    recall  f1-score   support

           Walking       1.00      1.00      1.00       537
  Walking Upstairs       0.63      0.98      0.77       491
Walking Downstairs       0.98      0.45      0.62       532
           Sitting       0.99      0.88      0.93       496
          Standing       0.74      1.00      0.85       420
            Laying       0.99      0.81      0.89       471

          accuracy                           0.85      2947
         macro avg       0.89      0.85      0.84      2947
      weighted avg       0.89      0.85      0.84      2947



## Logistic Regression hyperparameter tuning: Grid Search

In [1159]:
from sklearn.model_selection import RepeatedStratifiedKFold

solvers = ['newton-cg', 'lbfgs', 'liblinear']
penalty = ['l2']
c_values = [100, 10, 1.0, 0.1, 0.01]


logreg = LogisticRegression(random_state=16)

# define grid search
grid = dict(solver=solvers,penalty=penalty,C=c_values)
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

grid_search = GridSearchCV(estimator=logreg, param_grid=grid, n_jobs=-1, cv=cv, scoring='accuracy',error_score=0)
grid_result = grid_search.fit(X_train, y_train)

# summarize results
print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))

Best: 0.986534 using {'C': 10, 'penalty': 'l2', 'solver': 'liblinear'}


In [1171]:
grid_search.best_estimator_

LogisticRegression(C=10, random_state=16, solver='liblinear')

In [1173]:
print(classification_report(y_test,y_pred, target_names=target_names))

                    precision    recall  f1-score   support

           Walking       0.99      1.00      1.00       537
  Walking Upstairs       0.64      0.98      0.77       491
Walking Downstairs       0.98      0.49      0.65       532
           Sitting       0.98      0.91      0.94       496
          Standing       0.75      1.00      0.86       420
            Laying       1.00      0.78      0.88       471

          accuracy                           0.85      2947
         macro avg       0.89      0.86      0.85      2947
      weighted avg       0.89      0.85      0.85      2947



# SVM Classifier

In [1198]:
#Import svm model
from sklearn import svm

#Create a svm Classifier
#svmc = svm.SVC(kernel='linear', C=10) # Linear Kernel
svmc = svm.SVC(kernel='rbf') 
#svmc = svm.SVC(kernel="poly", degree=3) 

#Train the model using the training sets
svmc.fit(X_train, y_train)

#Predict the response for test dataset
y_pred = svmc.predict(X_test)

In [1205]:
# Model Accuracy: how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

print(classification_report(y_test,y_pred, target_names=target_names))                                         

Accuracy: 0.8659653885307091
                    precision    recall  f1-score   support

           Walking       1.00      1.00      1.00       537
  Walking Upstairs       0.69      0.98      0.81       491
Walking Downstairs       0.98      0.59      0.74       532
           Sitting       0.99      0.85      0.92       496
          Standing       0.72      1.00      0.84       420
            Laying       0.99      0.80      0.88       471

          accuracy                           0.87      2947
         macro avg       0.89      0.87      0.86      2947
      weighted avg       0.90      0.87      0.86      2947



## SVM hyperparameter tuning: Bayesian Optimisation

In [1208]:
# https://sites.gatech.edu/omscs7641/2024/02/16/tutorial-on-hyperparameter-tuning-using-scikit-learn/

from sklearn.model_selection import cross_val_score
from sklearn.metrics import classification_report
from GPyOpt.methods import BayesianOptimization
from skopt import BayesSearchCV, space, plots
from sklearn.svm import SVC

# Bayesian Search search space
search_spaces = {
    'C': (1e-2, 1e2, 'log-uniform'),
    'gamma': (1e-2, 1e2, 'log-uniform'),
    'kernel': ['rbf']
}

bayes_opt = BayesSearchCV(
    SVC(),
    search_spaces=search_spaces,
    n_iter=81,
    cv=5,
    n_jobs=4
)

bayes_opt.fit(X_train, y_train)
print(f'Best Params: {bayes_opt.best_params_}')

y_pred = bayes_opt.predict(X_test)
print(classification_report(y_test, y_pred))

Best Params: OrderedDict({'C': 10.250303216431895, 'gamma': 0.06465388719294979, 'kernel': 'rbf'})
                    precision    recall  f1-score   support

            LAYING       0.99      1.00      1.00       537
           SITTING       0.62      0.98      0.76       491
          STANDING       0.98      0.45      0.62       532
           WALKING       1.00      0.82      0.90       496
WALKING_DOWNSTAIRS       0.68      1.00      0.81       420
  WALKING_UPSTAIRS       1.00      0.77      0.87       471

          accuracy                           0.83      2947
         macro avg       0.88      0.84      0.83      2947
      weighted avg       0.89      0.83      0.83      2947



# KNN

In [1188]:
from sklearn.neighbors import KNeighborsClassifier

# output classes: 'LAYING' 'SITTING' 'STANDING' 'WALKING' 'WALKING_DOWNSTAIR', therefore k=6

knn = KNeighborsClassifier(n_neighbors=6)
knn.fit(X_train, y_train)

y_pred = knn.predict(X_test)

print("Accuracy:", metrics.accuracy_score(y_test, y_pred))

print(classification_report(y_test,y_pred, target_names=target_names))

Accuracy: 0.8883610451306413
                    precision    recall  f1-score   support

           Walking       1.00      1.00      1.00       537
  Walking Upstairs       0.80      0.94      0.87       491
Walking Downstairs       0.94      0.78      0.85       532
           Sitting       0.90      0.84      0.87       496
          Standing       0.75      0.93      0.83       420
            Laying       0.97      0.84      0.90       471

          accuracy                           0.89      2947
         macro avg       0.89      0.89      0.89      2947
      weighted avg       0.90      0.89      0.89      2947



# Ensemble Model

In [1326]:
# https://aryanbajaj13.medium.com/ensemble-models-how-to-make-better-predictions-by-combining-multiple-models-with-python-codes-6ac54403414e

from sklearn.ensemble import VotingClassifier
from sklearn.datasets import make_classification

# Define three different models
model1 = LogisticRegression(random_state=16)
model2 = KNeighborsClassifier(n_neighbors=6)
model3 = SVC(random_state=16, probability=True)
#model3 = SVC(random_state=16)
model4 = RandomForestClassifier(random_state=123)
model5 = DecisionTreeClassifier(criterion='entropy', max_depth=7, min_samples_leaf=20, random_state=16)

# Combine the models using majority voting
#ensemble = VotingClassifier(estimators=[('lr', model1), ('knn', model2), ('svc', model3), ('rf', model4), ('dt', model5)], voting='soft')
#ensemble = VotingClassifier(estimators=[('lr', model1), ('knn', model2), ('svc', model3), ('rf', model4)], voting='soft')
ensemble = VotingClassifier(estimators=[('lr', model1), ('knn', model2)], voting='soft')

In [1328]:
# Fit the ensemble on the training data
ensemble.fit(X_train, y_train)

VotingClassifier(estimators=[('lr', LogisticRegression(random_state=16)),
                             ('knn', KNeighborsClassifier(n_neighbors=6))],
                 voting='soft')

In [1329]:
# Evaluate the performance of the ensemble on the testing data
print(f"Accuracy of the ensemble: {ensemble.score(X_test, y_test)*100} %")

Accuracy of the ensemble: 88.63250763488293 %


# SUMMARY

##### Best model accuracies are summarised below.

Decision Tree (default hyperparameters): accuracy = 0.73 

Decision Tree (criterion='entropy', max_depth=7, min_samples_leaf=20,): accuracy = 0.75 

Random Forest (default hyperparameters): accuracy = 0.83

Logistic Regression (default hyperparameters): accuracy = 0.85

Support Vector Machine (default hyperparameters): accuracy = 0.85

KNN (k=6): accuracy = 0.89

# DISCUSSION

I have trained several models: DT, RF, LogReg, SVM, and KNN to classify human activities from the 
human-activity-recognition-with-smarthone dataset. The train dataset consists of 562 features and 7352 samples, while the test dataset is comprised of 2947 samples.

With such a large number of input features, I thought that dimentionality reduction might be beneficial.

I found that decision trees and random forests gave the best accuracies with models trained using the principal component datasets, while the remaining models (LogReg, SVM and KNN) gave better accuracies without dimensionality reduction.

Although my accuracy scores aren't competition winning, they are fairly good. Given more time, I would have liked to try neural network models and spent a bit more time fine tuning by best performing models.6